dftimewolf/lib/collectors/gce_disk_export.py

```diff
@@ -0,0 +1,103 @@
+# -*- coding: utf-8 -*-
+"""Export disk image from a GCP project to Google Cloud Storage."""
+
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import os
+from libcloudforensics import gcp
+
+from dftimewolf.lib import module
+from dftimewolf.lib.modules import manager as modules_manager
+
+
+class GoogleCloudDiskExport(module.BaseModule):
+  """Google Cloud Platform (GCP) Disk Export.
+
+  Attributes:
+    analysis_project (gcp.GoogleCloudProject): Project where the disk
+         image is created then exported.
+         If not exit, source_project will be used.
+    source_project (gcp.GoogleCloudProject): Source project
+        containing the disk to export.
+    source_disk (gcp.GoogleComputeDisk): Disk that needs to be exported.
+    gcs_output_location (str): Google Cloud Storage parent bucket/folder
+        path of the exported image.
+    exported_disk_name (str): Name of the output file, must comply
+        with ^[A-Za-z0-9-]*$' and '.tar.gz' will be appended to the name. If not
+        exist, random name will be generated.
+    _image_name (str): Intermediary image to export.
+  """
+
+  def __init__(self, state, critical=False):
+    """Initializes a Google Cloud Platform (GCP) Disk Export.
+
+    Args:
+      state (DFTimewolfState): recipe state.
+      critical (Optional[bool]): True if the module is critical, which causes
+          the entire recipe to fail if the module encounters an error.
+    """
+    super(GoogleCloudDiskExport, self).__init__(state, critical=critical)
+    self.analysis_project = None
+    self.source_project = None
+    self.source_disk = None
+    self.gcs_output_location = None
+    self.exported_disk_name = None
+    self._image_name = None
+
+  def Process(self):
+    """Creates and exports disk image to the output bucket."""
+    image_object = self.analysis_project.CreateImageFromDisk(
+        self.source_disk, name=self._image_name)
+    image_object.ExportImage(
+        self.gcs_output_location, output_name=self.exported_disk_name)
+    image_object.Delete()
+    output_uri = os.path.join(
+        self.gcs_output_location, '{0:s}.tar.gz'.format(
+            self.exported_disk_name))
+    print('Disk was exported to: {0:s}'.format(output_uri))
+    self.state.output.append((output_uri))
+
+  # pylint: disable=arguments-differ
+  def SetUp(self,
+            analysis_project_name,
+            source_project_name,
+            source_disk_name,
+            gcs_output_location,
+            exported_disk_name):
+    """Sets up a Google Cloud Platform (GCP) Disk Export.
+
+    This method creates the required objects to initialize
+    the GoogleCloudDiskExport class attributes.
+
+    If the analysis_project_name is not specified it will use the
+    source_project_name instead.
+
+    Args:
+      analysis_project_name (Optional[str]): Project ID where the
```

Comment:
This is not really true - `SetUp` will fail if no argument is provided for analysis_project_name (it's not a keyword argument). Like in the recipe, please put all optional parameters at the end and if they're optional, please add a default value in python (regardless of whether a default value is setup in the recipe parameters).

Reply:
There is a way to test IAM permissions via projects.testIamPermissions under the resource-manager APIs. This will need direct calls to Cloud APIs and I was thinking to add these to libcloudforensics, before the time consuming  operations. On the other side, the export function will trigger a cloud build job which will fail right away if it doesn't have access to the output bucket, it doesn't wait till the export is done to check. So I think it will save us little time here only if we check the permissions on the output bucket before the createImage function starts to create an image, what do you think?